### PR TITLE
[common] Added new unfused softmax cuda kernel to support causal attention mask with arbitrary query sequence length    

### PR DESCRIPTION
**Pull request by Oleg-Goncharov**
In this PR, a new unfused softmax cuda kernel was added to support causal attention mask with arbitrary query sequence length that enables speculative decoding. Max sequence length supported by this and other unfused softmax kernels is 16384.
 
Performance comparison (duration) of the new kernel vs. unfused upper triangular softmax run on L40 GPU for different sequence lengths:

Performance comparison

### CODE REVIEW

**Comment by cyanguwa** [comment]
I think the title could be a bit shorter :)

**Comment by ksivaman** [comment]
@Oleg-Goncharov could you sign-off your commits? The instructions are in [this guide](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst#sign-your-work).

**Comment by ptrendx** [comment]
@Oleg-Goncharov could you clean the git history? The PR shows some unrelated commits

**Review by timmoon10 — CHANGES_REQUESTED** [review]
Overall looks good. I have some suggestions, but I'm not too concerned since the problems are also present in [scaled_upper_triang_masked_softmax.cu](https://github.com/NVIDIA/TransformerEngine/blob/b5e13a16611be162538f489f3fd7096518640e15/transformer_engine/common/fused_softmax/scaled_upper_triang_masked_softmax.cu). My major request is to add a test, preferably a C++ test like https://github.com/NVIDIA/TransformerEngine/blob/main/tests/cpp/operator/test_transpose.cu. You could alternatively write a Python test in [tests/pytorch/test_numerics.py](https://github.com/NVIDIA/TransformerEngine/blob/b5e13a16611be162538f489f3fd7096518640e15/tests/pytorch/test_numerics.py).
